src/group_theory/monoid_localization.lean

```diff
@@ -266,25 +287,335 @@ show _ * _ = _, by rw [mul_inv_left, mul_one]
+    to `g x * (g y)⁻¹`, where `(x, y) : M × S` are such that `z = f x * (f y)⁻¹`. -/
+@[to_additive "Given a localization map `f : M →+ N` and a map of `add_comm_monoid`s `g : M →+ P` such that `g y` is invertible for all `y : S`, the homomorphism induced from `N` to `P` sending `z : N` to `g x - g y`, where `(x, y) : M × S` are such that `z = f x - f y`."]
+noncomputable def lift : N →* P :=
+{ to_fun := λ z, g (sec S f.1 z).1 * ↑(is_unit.lift_right (g.restrict S) hg (sec S f.1 z).2)⁻¹,
+  map_one' := by rw [mul_inv_left, mul_one]; exact f.eq_of_eq hg (by rw [←sec_spec f.3, one_mul]),
+  map_mul' := λ x y, by rw [mul_inv_left hg, ←mul_assoc, ←mul_assoc, mul_inv_right hg,
+    mul_comm _ (g (sec S f.1 y).1), ←mul_assoc, ←mul_assoc, mul_inv_right hg];
+    repeat { rw ←g.map_mul };
+    exact f.eq_of_eq hg (by repeat { rw f.1.map_mul <|> rw sec_spec' f.3 }; ac_refl) }
+
+variables {S g}
+
+@[to_additive] lemma lift_mk' (x y) :
+  f.lift hg (f.mk' x y) = g x * ↑(is_unit.lift_right (g.restrict S) hg y)⁻¹ :=
+(mul_inv hg).2 $ f.eq_of_eq hg $ by
+  rw [f.1.map_mul, f.1.map_mul, sec_spec' f.3, mul_assoc, f.mk'_spec, mul_comm]
+
+@[to_additive] lemma lift_spec (z v) :
+  f.lift hg z = v ↔ g (sec S f.1 z).1 = g (sec S f.1 z).2 * v :=
+mul_inv_left hg _ _ v
+
+@[to_additive] lemma lift_spec_mul (z w v) :
+  f.lift hg z * w = v ↔ g (sec S f.1 z).1 * w = g (sec S f.1 z).2 * v :=
+begin
+  rw mul_comm,
+  show _ * (_ * _) = _ ↔ _,
+  rw [←mul_assoc, mul_inv_left hg, mul_comm],
+end
+
+@[to_additive] lemma lift_mk'_spec (x v) (y : S) :
+  f.lift hg (f.mk' x y) = v ↔ g x = g y * v :=
+by rw f.lift_mk' hg; exact mul_inv_left hg _ _ _
+
+@[to_additive] lemma lift_mul_right (z) :
+  f.lift hg z * g (sec S f.1 z).2 = g (sec S f.1 z).1 :=
+show _ * _ * _ = _, by erw [mul_assoc, is_unit.lift_right_inv_mul, mul_one]
+
+@[to_additive] lemma lift_mul_left (z) :
+  g (sec S f.1 z).2 * f.lift hg z = g (sec S f.1 z).1 :=
+by rw [mul_comm, lift_mul_right]
+
+@[to_additive] lemma lift_of (x : M) :
+  f.lift hg (f.1 x) = g x :=
+by rw [lift_spec, ←g.map_mul]; exact f.eq_of_eq hg (by rw [sec_spec' f.3, f.1.map_mul])
+
+@[to_additive] lemma lift_eq_iff {x y : M × S} :
+  f.lift hg (f.mk' x.1 x.2) = f.lift hg (f.mk' y.1 y.2) ↔ g (x.1 * y.2) = g (y.1 * x.2) :=
+by rw [lift_mk', lift_mk', mul_inv hg]
+
+@[simp, to_additive] lemma  lift_comp_of : (f.lift hg).comp f.1 = g :=
+by ext; exact f.lift_of hg _
+
+@[to_additive] lemma is_unit_comp (j : N →* P) (y : S) :
+  is_unit (j.comp f.1 y) :=
+⟨units.map j $ is_unit.lift_right (f.1.restrict S) f.2 y,
+  show j _ = j _, from congr_arg j $ is_unit.coe_lift_right (f.1.restrict S) f.2 _⟩
+
+@[simp, to_additive] lemma lift_apply_of (j : N →* P) :
+  f.lift (f.is_unit_comp j) = j :=
+begin
+  ext,
+  rw lift_spec,
+  show j _ = j _ * _,
+  erw [←j.map_mul, sec_spec' f.3],
+end
+
+@[to_additive] lemma funext {j k : N →* P}
+  (h : ∀ a, j.comp f.1 a = k.comp f.1 a) : j = k :=
```

Comment:
This is the statement that localization maps are `epic` right? That might be a better name.